### ISSUE TITLE
When systemd.network has [Match] section with invalid entries, it is treated as empty section

### ISSUE THREAD

**Issue by WGH-** [systemd/systemd#4454]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

systemd 231  
+PAM -AUDIT -SELINUX +IMA -APPARMOR +SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL -XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN
### Used distribution

Gentoo Linux
### In case of bug report: Expected behaviour you didn't see

I create a .network file, and specified a rather long name in `[Match]` `Name=` property.

```
[Match]
Name=fffffffffffffffffffffffffffffff

[Network]
Address=192.168.0.1/24
```
### In case of bug report: Unexpected behaviour you saw

`systemd-networkd` logged an error saying "Interface name is not valid or too long, ignoring", and applied the address to all interfaces in the system, as if `[Match]` section was empty.

**Comment by poettering** [comment]
Yeah, a network file with an empty [Match] section will apply to all interfaces. And yes, we usuall will skip over invalid configuration settings and log abou them instead of failing entirely. 

I can see how this might be problematic in some cases. Not sure what we can do about this though.

**Comment by keszybz** [comment]
IIRC, I wanted empty [match] to match nothing. That'd come in handy here, even though it wouldn't solve the case where one of two matches triggered a syntax error (and so on).

Maybe we should treat all errors in match]

**Comment by keszybz** [comment]
... in [match] as fatal, since partial matches are probably almost always wrong.

**Comment by lucab** [comment]
This got reported by an user again here: https://github.com/coreos/bugs/issues/2015.

> A network config file with only a blank MATCH condition is then applied to all interfaces.

Whatever the result of this discussion is, it seems systemd should be fixed to either:
 * document the valid-empty-matches-all behavior
 * turn an empty field into a no-match, and document that case
 * turn an empty field into an error, and document that case

I'll be happy to take a look at any of those, as long as there is some consesus over the expected behavior. I'm personally in favor of considering the empty list an error and rejecting it.

/cc @soggiest

**Comment by amishmm** [comment]
I have a similar but slightly different bug report, but completely related to this so I am not creating another bug report.

While doing some testing I created a sample `.network` file (with highest priority).

After testing the same, I did not have immediate use of the test config but wanted to keep it as reference for future. So what I did is comment all the lines of it. (So effectively file became empty file with no section). i.e. it doesnt have `[Match]` section at all, so no question of invalid settings arises either.

What happened next is when today I restarted my system. All the interfaces remained unconfigured because the empty network file (having highest priority) matched everything.

This may be by design but in my opinion it is a bug. The logs do show this warning though:
`No valid settings found in the [Match] section. The file will match all interfaces`

My suggestion is that if file does not have any `[Match]` section OR if it is effectively empty then file should be ignored instead of being applied to all interfaces.

If anyone actually wants a config to match all interfaces they can put:
```
[Match]
Name=*
```

I can create separate bug report if required.
